dataRetriever/dataPool/headersCache/headersPool_test.go

```diff
@@ -0,0 +1,520 @@
+
+	header, err = headersCacher.GetHeaderByHash(headersHash[2012])
+	require.Error(t, headersCache.ErrHeaderNotFound, err)
+}
+
+func TestHeadersCacher_AddHeadersWithDifferentShardIdOnMultipleGoroutines(t *testing.T) {
+	t.Parallel()
+
+	cacheSize := 1001
+	numHdrsToGenerate := 1000
+
+	headersShard0, hashesShad0 := createASliceOfHeadersNonce0(numHdrsToGenerate, 0)
+	headersShard1, hashesShad1 := createASliceOfHeaders(numHdrsToGenerate, 1)
+	headersShard2, hashesShad2 := createASliceOfHeaders(numHdrsToGenerate, 2)
+	numElemsToRemove := 500
+
+	headersCacher, _ := headersCache.NewHeadersCacher(cacheSize, numElemsToRemove)
+
+	var waitgroup sync.WaitGroup
+	start := time.Now()
+	for i := 0; i < numHdrsToGenerate; i++ {
+		waitgroup.Add(5)
+		go func(index int) {
+			headersCacher.Add(hashesShad0[index], &headersShard0[index])
+			waitgroup.Done()
+		}(i)
+
+		go func(index int) {
+			headersCacher.Add(hashesShad1[index], &headersShard1[index])
+			go func(index int) {
+				headersCacher.RemoveHeaderByHash(hashesShad1[index])
+				waitgroup.Done()
+			}(index)
+			waitgroup.Done()
+		}(i)
+
+		go func(index int) {
+			headersCacher.Add(hashesShad2[index], &headersShard2[index])
+			go func(index int) {
+				headersCacher.RemoveHeaderByHash(hashesShad2[index])
+				waitgroup.Done()
+			}(index)
+			waitgroup.Done()
+		}(i)
+	}
+
+	waitgroup.Wait()
+
+	for i := 0; i < numHdrsToGenerate; i++ {
+		waitgroup.Add(1)
+		go func(index int) {
+			headersCacher.RemoveHeaderByHash(hashesShad0[index])
+			waitgroup.Done()
+		}(i)
+	}
+	waitgroup.Wait()
+
+	elapsed := time.Since(start)
+	fmt.Printf("time need to add %d in cache %s \n", numHdrsToGenerate, elapsed)
+
+	require.Equal(t, 0, headersCacher.GetNumHeadersFromCacheShard(0))
+	require.Equal(t, 0, headersCacher.GetNumHeadersFromCacheShard(1))
+	require.Equal(t, 0, headersCacher.GetNumHeadersFromCacheShard(2))
+}
+
+func TestHeadersCacher_TestEvictionRemoveCorrectHeader(t *testing.T) {
+	t.Parallel()
+
+	shardId := uint32(0)
+	cacheSize := 2
+	numHeadersToGenerate := 3
+
+	headers, headersHashes := createASliceOfHeaders(numHeadersToGenerate, shardId)
+	headersCacher, _ := headersCache.NewHeadersCacher(cacheSize, 1)
+
+	for i := 0; i < numHeadersToGenerate-1; i++ {
+		headersCacher.Add(headersHashes[i], &headers[i])
+		time.Sleep(time.Microsecond)
+	}
+
+	header, err := headersCacher.GetHeaderByHash(headersHashes[0])
+	require.Nil(t, err)
+	require.Equal(t, &headers[0], header)
+
+	headersCacher.Add(headersHashes[2], &headers[2])
+
+	header, err = headersCacher.GetHeaderByHash(headersHashes[0])
+	require.Nil(t, err)
+	require.Equal(t, &headers[0], header)
+
+	header, err = headersCacher.GetHeaderByHash(headersHashes[2])
+	require.Nil(t, err)
+	require.Equal(t, &headers[2], header)
+
+	header, err = headersCacher.GetHeaderByHash(headersHashes[1])
+	require.Equal(t, headersCache.ErrHeaderNotFound, err)
+}
+
+func TestHeadersPool_AddHeadersMultipleShardsShouldWork(t *testing.T) {
```

Comment:
`ShouldWork` is not really necessary. Same for other tests.

Reply:
done.